components/spiffs/esp_spiffs.c

```diff
@@ -706,17 +757,36 @@ static int vfs_spiffs_readdir_r(void* ctx, DIR* pdir, struct dirent* entry,
         return errno;
     }
     const char * item_name = (const char *)out.name;
+
+    // Get file stat, used for setting file type in dirend entry
+    spiffs_stat s = {0};
+    off_t ret = SPIFFS_stat(efs->fs, item_name, &s);
+    if (ret < 0) {
+        errno = spiffs_res_to_errno(SPIFFS_errno(efs->fs));
+        SPIFFS_clearerr(efs->fs);
+        return errno;
+    }
+
     size_t plen = strlen(dir->path);
     if (plen > 1) {
-        if (strncasecmp(dir->path, (const char *)out.name, plen) || out.name[plen] != '/' || !out.name[plen+1]) {
+        // on subdirectory
+        if ((strncasecmp((const char *)out.name, dir->path, plen)) || (strncasecmp(dir->path, (const char *)out.name, plen) || out.name[plen] != '/' || !out.name[plen+1])) {
             return vfs_spiffs_readdir_r(ctx, pdir, entry, out_dirent);
         }
         item_name += plen + 1;
-    } else if (item_name[0] == '/') {
-        item_name++;
+    }
+    else {
+        // on root
+        if ((strlen(item_name) > 2) && (strchr(item_name+1, '/'))) {
+            return vfs_spiffs_readdir_r(ctx, pdir, entry, out_dirent);
+        }
+        if (item_name[0] == '/') {
+            item_name++;
+        }
     }
     entry->d_ino = 0;
-    entry->d_type = out.type;
+    if (s.meta[CONFIG_SPIFFS_META_LENGTH] == SPIFFS_TYPE_DIR) entry->d_type = DT_DIR;
```

Comment:
Suggest defining a structure for SPIFFS metadata and accessing fields of this structure, rather than bytes at certain offsets. This will be more robust against future changes to metadata format.